prov/shm/src/smr_msg.c

```diff
@@ -312,22 +311,24 @@ static ssize_t smr_generic_sendmsg(struct smr_ep *ep, const struct iovec *iov,
 	}
 
 	ofi_spin_lock(&ep->tx_lock);
-	iface = smr_get_mr_hmem_iface(ep->util_ep.domain, desc, &device);
 
 	total_len = ofi_total_iov_len(iov, iov_count);
 	assert(!(op_flags & FI_INJECT) || total_len <= SMR_INJECT_SIZE);
 
 	/* Do not inline/inject if IPC is available so device to device
 	 * transfer may occur if possible. */
-	use_ipc = ofi_hmem_is_ipc_enabled(iface) && (iov_count == 1) &&
-		  desc && (smr_get_mr_flags(desc) & FI_HMEM_DEVICE_ONLY) &&
-		  !(op_flags & FI_INJECT);
-
-	proto = smr_select_proto(use_ipc, smr_cma_enabled(ep, peer_smr), iface,
-				 op, total_len, op_flags);
+	if (iov_count == 1 && desc && desc[0]) {
```

Comment:
Could you explain what's magical about `desc[0]`?

In the current revision, if `iov_count  = 2` and `desc[1]` actually carries the cuda MR, cuda ipc won't be activated at all. Did I miss anything?

Reply:
That is the old implementation too. in order to use ipc you cannot send with multiple iovs. It is more complicated to use IPC for different ifaces.

Reply:
Right, multiple iovs have to take the SAR path. IPC currently does not support multiple iovs/ifaces. The biggest limitation is having to send multiple IPC handles in the command which previously could only hold one IPC handle. Recently, we doubled the command size to 256 bytes so we could theoretically hold more handles in the command. It will still get complicated though to have mixed iovs for IPC because the system memory would not have an IPC handle. We could figure out a way to send the virtual addresses to use CMA as the "IPC" method but that would still be complicated and also wouldn't cover the non-CMA cases so then we would have to take SAR. But that leads down a hole that would require a lot of modification to support... I'm not sure how worth it that is. Do you think that's something worth while to investigate from your point of view?

Reply:
@wenduwan @aingerson I do not think we really need that feature. Currently EFA provider only send msg with multiple (2) iovs to shm ep for host memory in EAGER protocol. The first iov stores additional packet headers for efa provider. The 2nd iov is the real msg iov for user buffer. This extra iov can be removed after we sending msg via shm directly in the future after onboarding peer API. 

And currently we have observed that for cuda-to-cuda latency, doing hmem copy inside efa, and use shm for pure host-to-host  has a better performance for small message sizes than making shm do IPC for cuda-to-cuda. I guess an improvement for shm in the future could be making such ipc/non-ipc switch point configurable in shm provider.

Reply:
> Do you think that's something worth while to investigate from your point of view?

Thanks for both's explanation. I'm not familiar with IPC(especially inside prov/shm) and from what I'm reading, the IPC path is rather a corner case. I'm biased toward prioritizing the crash as mentioned in https://github.com/ofiwg/libfabric/issues/6639.

Reply:
@shijin-aws I have thought about configuring that threshold for IPC but haven't done any experimentation to determine where that would be. I'm totally open to changing it though. We can easily do that adjustment either with the use_ipc definition or inside of select_proto. Feel free to experiment with it and open a PR for the change